src/Oscoin/Crypto/Blockchain/Genesis.hs

```diff
@@ -0,0 +1,134 @@
+import qualified Oscoin.Time as Time
+
+import           Codec.Serialise
+import           Control.Monad.Fail
+import           Data.ByteArray (ByteArrayAccess)
+
+
+-- | Set of parameters that uniquely define the genesis block.
+data GenesisParameters c = GenesisParameters
+    { gpBeneficiary     :: Beneficiary c
+    , gpTimestamp       :: Timestamp
+    , gpStartDifficulty :: Difficulty
+    , gpPoW             :: PoW
+    } deriving (Generic)
+
+deriving instance (Eq (Beneficiary c)) => Eq (GenesisParameters c)
+deriving instance (Show (Beneficiary c)) => Show (GenesisParameters c)
+
+
+instance (Aeson.ToJSON (Crypto.ShortHash c)) => Aeson.ToJSON (GenesisParameters c) where
+    toJSON GenesisParameters{..} = Aeson.object
+        [ "timestamp" .= Time.sinceEpoch gpTimestamp
+        , "startDifficulty" .= prettyDifficulty gpStartDifficulty
+        , "beneficiary" .= gpBeneficiary
+        , "pow" .= nonce
+        ]
+      where
+        PoW nonce = gpPoW
+
+instance (Aeson.FromJSON (Crypto.ShortHash c)) => Aeson.FromJSON (GenesisParameters c) where
+    parseJSON = Aeson.withObject "GenesisParameters" $ \o -> do
+        gpTimestamp <- Time.fromEpoch <$> o .: "timestamp"
+        gpStartDifficulty <- parseJsonDifficulty =<< o .: "startDifficulty"
+        gpBeneficiary <- o .: "beneficiary"
+        gpPoW <- PoW <$> o .: "pow"
+        pure $ GenesisParameters{..}
+      where
+        parseJsonDifficulty =
+            Aeson.withText "Difficulty" $ \t ->
+                case parseDifficulty t of
+                    Just d  -> pure d
+                    Nothing -> fail "Error decoding difficulty"
+
+
+-- | Build a genesis block from the given parameters. The block is
+-- intrinsically valid. It’s state hash references the 'mempty'
+-- 'LegacyTxState'
+buildGenesisBlock
+    :: forall c tx. ( Serialise (Crypto.Hash c)
+       , Crypto.HasHashing c
+       , Serialise (Beneficiary c)
+       , AuthTree.MerkleHash (Crypto.Hash c)
+       )
+    => GenesisParameters c -> Block c tx (Sealed c PoW)
+buildGenesisBlock gp = mkBlock blockHeader (gpBeneficiary gp) []
+  where
+    blockHeader = BlockHeader
+        { blockDataHash = noTxDataHash (gpBeneficiary gp)
+        , blockStateHash = initialStateHash
+        , blockTimestamp = gpTimestamp gp
+        , blockHeight = 0
+        , blockPrevHash = Crypto.zeroHash
+        , blockTargetDifficulty = gpStartDifficulty gp
+        , blockSeal = SealedWith (gpPoW gp)
+        }
+
+-- | Create 'GenesisParameters' for the given data. Computes the proof
+-- of work so that the block generated
+createGenesisParameters
+    :: ( Serialise (Crypto.Hash c)
+       , Crypto.HasHashing c
+       , ByteArrayAccess (BlockHash c)
+       , AuthTree.MerkleHash (Crypto.Hash c)
+       , Serialise (Beneficiary c)
+       ) => Beneficiary c -> Timestamp -> Difficulty -> Maybe (GenesisParameters c)
+createGenesisParameters gpBeneficiary gpTimestamp gpStartDifficulty = do
+    gpPoW <- findPoW blockHeader
```

Comment:
Kind of odd that creating the parameters involves PoW - I think it would make more sense if this was passed in.